app/forms/delivery_partner_form.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+class DeliveryPartnerForm
+  include ActiveModel::Model
+
+  attr_accessor :name, :lead_providers, :provider_relationships
+  validate :lead_providers_and_cohorts_validation
+
+  def available_lead_providers
+    LeadProvider.includes(:cohorts).select { |lead_provider| lead_provider.cohorts.any? }
```

Comment:
Is this just LeadProvider.joins(:cohorts) ?

Reply:
Nope, these result in different queries.
includes gives:
```
LeadProvider Load (0.4ms)  SELECT "lead_providers".* FROM "lead_providers"
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  HABTM_Cohorts Load (0.7ms)  SELECT "cohorts_lead_providers".* FROM "cohorts_lead_providers" WHERE "cohorts_lead_providers"."lead_provider_id" IN ($1, $2, $3)  [[nil, "25a1f5b2-e7a2-4563-8c71-f9b346140cf6"], [nil, "976e5988-bea5-4498-805c-1f8c5be3b8da"], [nil, "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"]]
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  Cohort Load (0.4ms)  SELECT "cohorts".* FROM "cohorts" WHERE "cohorts"."id" IN ($1, $2)  [[nil, "a44f7145-9df0-4905-9966-37950e507d8f"], [nil, "2d9cb19d-2ac9-4ec0-9a0f-41f30bffaa15"]]
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  CACHE LeadProvider Load (0.0ms)  SELECT "lead_providers".* FROM "lead_providers"
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  CACHE HABTM_Cohorts Load (0.0ms)  SELECT "cohorts_lead_providers".* FROM "cohorts_lead_providers" WHERE "cohorts_lead_providers"."lead_provider_id" IN ($1, $2, $3)  [[nil, "25a1f5b2-e7a2-4563-8c71-f9b346140cf6"], [nil, "976e5988-bea5-4498-805c-1f8c5be3b8da"], [nil, "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"]]
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" WHERE "cohorts"."id" IN ($1, $2)  [[nil, "a44f7145-9df0-4905-9966-37950e507d8f"], [nil, "2d9cb19d-2ac9-4ec0-9a0f-41f30bffaa15"]]
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
```

joins gives:
```
  LeadProvider Load (0.5ms)  SELECT "lead_providers".* FROM "lead_providers" INNER JOIN "cohorts_lead_providers" ON "cohorts_lead_providers"."lead_provider_id" = "lead_providers"."id" INNER JOIN "cohorts" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id"
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  Cohort Exists? (0.5ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  CACHE Cohort Exists? (0.0ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  Cohort Exists? (0.5ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  CACHE Cohort Exists? (0.0ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  Cohort Load (0.5ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  Cohort Load (0.6ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  CACHE LeadProvider Load (0.0ms)  SELECT "lead_providers".* FROM "lead_providers" INNER JOIN "cohorts_lead_providers" ON "cohorts_lead_providers"."lead_provider_id" = "lead_providers"."id" INNER JOIN "cohorts" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id"
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  CACHE Cohort Exists? (0.0ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  CACHE Cohort Exists? (0.0ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  CACHE Cohort Exists? (0.0ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  CACHE Cohort Exists? (0.0ms)  SELECT 1 AS one FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1 LIMIT $2  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"], ["LIMIT", 1]]
  ↳ app/forms/delivery_partner_form.rb:10:in `block in available_lead_providers'
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "976e5988-bea5-4498-805c-1f8c5be3b8da"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
  CACHE Cohort Load (0.0ms)  SELECT "cohorts".* FROM "cohorts" INNER JOIN "cohorts_lead_providers" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id" WHERE "cohorts_lead_providers"."lead_provider_id" = $1  [["lead_provider_id", "4307c0f6-c5a4-4cf0-a33b-b3b416db2032"]]
  ↳ app/views/admin/delivery_partners/choose_lead_providers.html.erb:21
```

Reply:
I think the only difference is the include so LeadProvider.joins(:cohorts).include(:cohorts) should work. It would fetch the right records from the db rather than using slow ruby to do the filtering.

Reply:
Yep, much better 
```
SQL (5.6ms)  SELECT "lead_providers"."id" AS t0_r0, "lead_providers"."created_at" AS t0_r1, "lead_providers"."updated_at" AS t0_r2, "lead_providers"."name" AS t0_r3, "cohorts"."id" AS t1_r0, "cohorts"."created_at" AS t1_r1, "cohorts"."updated_at" AS t1_r2, "cohorts"."start_year" AS t1_r3 FROM "lead_providers" INNER JOIN "cohorts_lead_providers" ON "cohorts_lead_providers"."lead_provider_id" = "lead_providers"."id" INNER JOIN "cohorts" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id"
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
  CACHE SQL (0.0ms)  SELECT "lead_providers"."id" AS t0_r0, "lead_providers"."created_at" AS t0_r1, "lead_providers"."updated_at" AS t0_r2, "lead_providers"."name" AS t0_r3, "cohorts"."id" AS t1_r0, "cohorts"."created_at" AS t1_r1, "cohorts"."updated_at" AS t1_r2, "cohorts"."start_year" AS t1_r3 FROM "lead_providers" INNER JOIN "cohorts_lead_providers" ON "cohorts_lead_providers"."lead_provider_id" = "lead_providers"."id" INNER JOIN "cohorts" ON "cohorts"."id" = "cohorts_lead_providers"."cohort_id"
  ↳ app/forms/delivery_partner_form.rb:10:in `available_lead_providers'
```